### PR TITLE
feat: add type hints for public methods

### LINKED ISSUE #157: feat: add type hints to methods
Having typings in the library would help and saw that there is a PR opened for that. However, the documentation is somewhat lacking.

It is hard to tell what the methods return and are all the parameters while looking at the docs https://googleapis.dev/python/bigquery/latest/generated/google.cloud.bigquery.client.Client.html

For example, in the tutorial: what is the difference between these 2 table objects created?
```
table = bigquery.Table(table_id, schema=schema)
table = client.create_table(table)  # Make an API request.
```

### PR DESCRIPTION
**Pull request by plamut**
Closes #157.
Supersedes #445.

**PR checklist:**
- [x] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/python-bigquery/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [x] Ensure the tests and linter pass
- [x] Code coverage does not decrease (if any source code was changed)
- [x] Appropriate docs were updated (if necessary)

### CODE REVIEW

File: noxfile.py
Comment (tswast) [inline noxfile.py:95]: Can we add this to `nox.options.sessions` so that it runs in the `Kokoro` build?

**Comment by tswast** [comment]
Beautiful! I'm OOO until Tuesday, so I can't take a look at this too closely right now, but happy to get it merged assuming we get it running in Kokoro and someone else reviews.

Oh, and can we get `pytype` running in our `pre-commit` config?

**Comment by plamut** [comment]
> Oh, and can we get pytype running in our pre-commit config?

The check is too slow, unfortunately, it takes like ~90 seconds on my machine, which disrupts the workflow too much. Unless, of course, your experience is different and it's my machine that needs some tweaking.

I noticed that, `pytype` does seem to cache the results, but when I added a trivial change (a single blank line in `client.py`), running it once more again took the same as before.

File: noxfile.py
Comment (plamut) [reply noxfile.py:96]: Done. From [Kokoro log](https://source.cloud.google.com/results/invocations/fa869575-8829-4345-8cbc-ed10b88abed7/targets/cloud-devrel%2Fclient-libraries%2Fpython%2Fgoogleapis%2Fpython-bigquery%2Fpresubmit%2Fpresubmit/log):
```
...
nox > Session blacken was successful.
nox > Running session pytype
nox > Creating virtual environment (virtualenv) using python3.8 in .nox/pytype
nox > python -m pip install pytype==2021.4.9
nox > pytype
Computing dependencies
Analyzing 40 sources with 7 local dependencies
ninja: Entering directory `.pytype'
[1/69] check google.cloud.bigquery.encryption_configuration
...
```

File: google/cloud/bigquery/client.py
Comment (tswast) [inline google/cloud/bigquery/client.py:1664]: Any reason we don't have a return type here?

```suggestion
    ) -> _QueryResults:
```

File: google/cloud/bigquery/client.py
Comment (tswast) [inline google/cloud/bigquery/client.py:1722]: Can we do a union type here?

Alternatively, what do you think about introducing some `abc`-style module for all these different job classes? I know we have the [_AsyncJob](https://github.com/googleapis/python-bigquery/blob/6502a602337ae562652a20b20270949f2c9d5073/google/cloud/bigquery/job/base.py#L131) class, but that's private and also poorly named given that it's not usable in true `async` code.

File: google/cloud/bigquery/job/base.py
Comment (tswast) [inline google/cloud/bigquery/job/base.py:654]: Since some job classes do use this implementation without overriding it, I wonder if this is more evidence that we should make `_AsyncJob` public (just `Job`?) / make a separate `abc` type for jobs?

File: google/cloud/bigquery/job/base.py
Comment (tswast) [inline google/cloud/bigquery/job/base.py:855]: Maybe we need `abc` types for configurations too?

File: google/cloud/bigquery/job/query.py
Comment (tswast) [inline google/cloud/bigquery/job/query.py:1283]: Why `Any` and not `"pandas.DataFrame"`?

File: google/cloud/bigquery/job/query.py
Comment (tswast) [inline google/cloud/bigquery/job/query.py:1212]: Why `Any` and not `"pyarrow.Table"`?

File: google/cloud/bigquery/table.py
Comment (tswast) [inline google/cloud/bigquery/table.py:51]: As much as I think it'd be quite useful to depend on `google-cloud-bigquery-storage` being present, I don't think we wanted to make `google-cloud-bigquery-storage` a required dependency just for this feature.

How do other packages support cross-package types for their "extras"?

**Review by tswast — CHANGES_REQUESTED** [review]
Looking good!

I have a few questions about how we want to handle job classes, but if we make any changes beyond a few more `Unions` such as `abc` types, I think that warrants a separate PR.

Also curious about cross-package types. I could actually probably be convinced to make `google-cloud-bigquery-storage` a required dependency, but I don't want to do so by accident.

**Comment by tswast** [comment]
Looks like we can conditionally import, only if we are currently type checking: https://stackoverflow.com/a/57940134/101923

Not sure if `pytype` respects this or if it's just a `mypy` thing.

File: google/cloud/bigquery/job/query.py
Comment (plamut) [reply google/cloud/bigquery/job/query.py:1283]: Perhaps because of [this](https://github.com/googleapis/python-bigquery/pull/445#discussion_r549215027)? I'll nevertheless check again if a quick workaround can be found.

File: google/cloud/bigquery/table.py
Comment (plamut) [reply google/cloud/bigquery/table.py:51]: It appears pytype supports this as well, will use it.

https://github.com/google/pytype/blob/master/docs/faq.md#how-do-i-write-code-that-is-seen-by-pytype-but-ignored-at-runtime

**Comment by plamut** [comment]
> Looks like we can conditionally import, only if we are currently type checking: https://stackoverflow.com/a/57940134/101923
> 
> Not sure if `pytype` respects this or if it's just a `mypy` thing.

It appears pytype supports this as well, will use it.

https://github.com/google/pytype/blob/master/docs/faq.md#how-do-i-write-code-that-is-seen-by-pytype-but-ignored-at-runtime

File: google/cloud/bigquery/job/query.py
Comment (plamut) [reply google/cloud/bigquery/job/query.py:1283]: Good news - I figured out that if we install all dependencies in the nox session before running the type checks, A LOT of these import errors disappear, and we don't need to make workarounds such as this one or silencing the checker all around the code.

File: google/cloud/bigquery/client.py
Comment (plamut) [reply google/cloud/bigquery/client.py:1718]: I kind of dislike "wide" union hints that clutter the code, thus introducing a base class for these type sounds reasonable. But perhaps in a separate PR, since that's an actual code refactoring, and not merely annotations that don't affect the runtime?

File: google/cloud/bigquery/job/base.py
Comment (plamut) [reply google/cloud/bigquery/job/base.py:653]: If that's not something end users should know about, I'd prefer keeping it private, but I support the renaming idea - "async" is too often automatically associated with `async.io` nowadays.

Ditto for adding an `ABC`, if we decide for that with other class families, such as job configs.

**Comment by plamut** [comment]
Whoa, the very first test that I wrote in this library caught a method signature mismatch (incompatible annotations). :)

Will fix it.

**Comment by plamut** [comment]
I don't know why imports directly from `google.cloud` still cause false import errors, but fortunately there are not many of them. On the other hand, it was possible to get rid of the vast majority of those noisy "ignore import error" directives.

**Review by tswast — COMMENTED** [review]
Looks great!!!!

I'm holding off on approval because I'd like a unit test nox session implemented that doesn't have any "extras" so we can verify that at least minimal functionality is supported if pandas/pyarrow/google-cloud-bigquery-storage aren't installed https://github.com/googleapis/python-bigquery/issues/621

I trust that `if typing.TYPE_CHECKING:` is working as expected, but I'd like us to verify too

**Review by tswast — APPROVED** [review]
LGTM assuming Kokoro passes

**Comment by tswast** [comment]
Marking CLI as yes, as this PR is based on work done for hire last year.

**Pull request merged**
